### PR TITLE
Add support for dependencies in plugin descriptor properties with semver range 

### LINKED ISSUE #1707: [opensearch-plugin] Cannot install old patch version of plugins on newer opensearch builds
```
2021-12-10 21:54:54 INFO     Executing "/tmp/tmp3_3h_wbz/opensearch-1.2.1/bin/opensearch-plugin install --batch file:/tmp/tmp3_3h_wbz/opensearch-job-scheduler-1.2.0.0.zip" in /tmp/tmp3_3h_wbz/opensearch-1.2.1
-> Installing file:/tmp/tmp3_3h_wbz/opensearch-job-scheduler-1.2.0.0.zip
-> Downloading file:/tmp/tmp3_3h_wbz/opensearch-job-scheduler-1.2.0.0.zip
-> Failed installing file:/tmp/tmp3_3h_wbz/opensearch-job-scheduler-1.2.0.0.zip
-> Rolling back file:/tmp/tmp3_3h_wbz/opensearch-job-scheduler-1.2.0.0.zip
-> Rolled back file:/tmp/tmp3_3h_wbz/opensearch-job-scheduler-1.2.0.0.zip
Exception in thread "main" java.lang.IllegalArgumentException: Plugin [opensearch-job-scheduler] was built for OpenSearch version 1.2.0 but version 1.2.1 is running
	at org.opensearch.plugins.PluginsService.verifyCompatibility(PluginsService.java:390)
	at org.opensearch.plugins.InstallPluginCommand.loadPluginInfo(InstallPluginCommand.java:803)
	at org.opensearch.plugins.InstallPluginCommand.installPlugin(InstallPluginCommand.java:858)
	at org.opensearch.plugins.InstallPluginCommand.execute(InstallPluginCommand.java:263)
	at org.opensearch.plugins.InstallPluginCommand.execute(InstallPluginCommand.java:237)
	at org.opensearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:100)
	at org.opensearch.cli.Command.mainWithoutErrorHandling(Command.java:138)
	at org.opensearch.cli.MultiCommand.execute(MultiCommand.java:104)
	at org.opensearch.cli.Command.mainWithoutErrorHandling(Command.java:138)
	at org.opensearch.cli.Command.main(Command.java:101)
	at org.opensearch.plugins.PluginCli.main(PluginCli.java:60)
```

This is blocking us from consuming the redistribution artifacts for patch releases.

### PR DESCRIPTION
**Pull request by abseth-amzn**
### Description
As a first step towards supporting semVer range of compatible versions for plugins, we are starting with support for specifying 'dependencies' for a plugin in its plugin-descriptor.properties file. The change currently supports specification of only one dependency (opensearch). Following notations are supported for specifying the compatible dependency version:
* dependencies={ opensearch: x.y.z }
* dependencies={ opensearch: =x.y.z }
* dependencies={ opensearch: ~x.y.z } 
(specification of major, minor and patch version is required in above notations)

### Related Issues
- Resolves #1707 

### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [ ] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [X] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by dblock — APPROVED** [review]
I like this very much! Iterate to 🍏 .

Is there really no semver parser library that we can bring in here instead of rolling our own? I understand OpenSearch wants to minimize dependencies, but maybe this is not the place to do that? Curious what others think?

There's likely developer documentation and other .md's in this repo that need to be updated. 

We also probably need integration tests, and bcw tests for upgrade scenarios.

Maybe someone else can do a thorough pass, @peternied?

**Comment by abseth-amzn** [comment]
Thanks @dblock.
I have removed unused gson licenses so that precommit check succeeds.

I had looked into few semver java libraries:
1. https://github.com/vdurmont/semver4j - no longer maintained.
2. https://github.com/zafarkhaja/jsemver/ - has a major version of zero with a note saying its public API should not be considered stable.
3. https://github.com/semver4j/semver4j - seemed like a viable candidate. However it uses a Version type ("1.2.3-beta.4+sha899d8g79f87") inherently in its library that looks little different than Version type used by Opensearch. The functionality we need looked simple enough to add. Looking for thoughts if using this library (or any other) is preferred. 

Adding >=X.Y.Z would be simple, but I thought it would go well with supporting multiple ranges [>=X.Y.Z, <A.B.C] so that range is bounded. Adding support for multiple ranges could be a follow up to this PR?

File: server/src/main/java/org/opensearch/plugins/PluginInfo.java
Comment (peternied) [reply server/src/main/java/org/opensearch/plugins/PluginInfo.java:74]: Or do you think it would be better to follow the example DB suggested of `>=X.Y.Z`?

**Review by peternied — COMMENTED** [review]
Great showing @abseth-amzn - I suspect a couple of plugin owners might want to buy you a 🍺 of your choice after this is merged.

A number of comments inline, feel free to dismiss the nits without comment if you feel otherwise.

File: server/src/main/java/org/opensearch/plugins/PluginInfo.java
Comment (abseth-amzn) [reply server/src/main/java/org/opensearch/plugins/PluginInfo.java:276]: Just to begin with, adding support for one dependency and few range notations. We can iterate on this and add support for more dependencies and notations going forward.

File: server/src/main/java/org/opensearch/plugins/PluginInfo.java
Comment (abseth-amzn) [reply server/src/main/java/org/opensearch/plugins/PluginInfo.java:74]: Even in the case of >=X.Y.Z, it would be better to have support for multiple ranges to accommodate [>=X.Y.Z, <A.B.C] so that range can be bounded. I can change this to a list.

File: server/build.gradle
Comment (peternied) [reply server/build.gradle:151]: To expand on this please look into the compatibility logs for what failed when they attempted to build, it might be as easy as a version number update on gson or it might be more complex.

image

#### Links to sections of that doc:

- SQL https://github.com/opensearch-project/OpenSearch/actions/runs/7100520996/job/19326777156?pr=11441#step:4:14126
- Reporting https://github.com/opensearch-project/OpenSearch/actions/runs/7100520996/job/19326777156?pr=11441#step:4:12319

Comment (abseth-amzn) [reply server/build.gradle:151]: I bumped the gson version in reporting and sql projects locally to 2.10.1 and they seem to be building fine with ./gradlew assemble. Will follow up with a PR for those projects. Not sure if there might be compatibility check failures (downstream build break) for them though.

File: server/src/main/java/org/opensearch/plugins/PluginInfo.java
Comment (peternied) [inline server/src/main/java/org/opensearch/plugins/PluginInfo.java:191]: How will we handle this in older versions, I'm concerned that PluginInfo serialized on `v2.N-1` won't be compatible with `v2.N`.  I think the only want to handle this during the backport is to support both the existing value and a new value.

Comment (peternied) [reply server/src/main/java/org/opensearch/plugins/PluginInfo.java:191]: Note; this is OK on v3.0.0 since we can introduce breaking changes, but we will need to add a deprecated noticed on the field in a 2.X release.

File: server/src/main/java/org/opensearch/plugins/PluginsService.java
Comment (peternied) [inline server/src/main/java/org/opensearch/plugins/PluginsService.java:405]: I don't think the logic of this function is correct, I think you need to iterate through all of the ranges and if any match, then you can return true, otherwise return false at line @412.

Comment (abseth-amzn) [reply server/src/main/java/org/opensearch/plugins/PluginsService.java:405]: When plugin authors specify multiple ranges, lets say [>=A.B.C, <X.Y.Z] then the core should match its engine version against all the ranges and should allow only if it matches all of them. This would help make the range bounded. Otherwise, [>=A.B.C, <X.Y.Z]  is as good as [>=A.B.C].
I found few examples of this behavior here - https://guides.rubygems.org/patterns/.
Thoughts?

Comment (peternied) [reply server/src/main/java/org/opensearch/plugins/PluginsService.java:405]: With this PR there are only 2 operators `~` `=`, how can you say a plugin works with both versions 2.0.X and 2.1.X, wouldn't that be `[~2.0.0, ~2.1.0]`?

Comment (abseth-amzn) [reply server/src/main/java/org/opensearch/plugins/PluginsService.java:405]: I see. As support for more operators [ <, >, !, <=, >= ] is added, [~2.0.0, ~2.1.0] can be achieved with [ >=2.0.0, < 2.2.0 ]. However, if we don't enforce satisfying all ranges, then supporting bounded range with [<, >] operators would not be possible.

Comment (peternied) [reply server/src/main/java/org/opensearch/plugins/PluginsService.java:405]: @abseth-amzn It sounds like we might need to invest more in this space to support this scenario.  I think its a great benefit to allow plugins to state capability with patch versions - as I imagine many would adopt this new setting immediately.

To unblock this, can you validate that number of values in the version array is one to avoid any immediate backward compatibility issues and create a follow up issue the encompasses this scenario?

File: server/src/main/java/org/opensearch/plugins/PluginInfo.java
Comment (peternied) [reply server/src/main/java/org/opensearch/plugins/PluginInfo.java:191]: Any thoughts on addressing this?

File: server/src/main/java/org/opensearch/plugins/PluginsService.java
Comment (dblock) [reply server/src/main/java/org/opensearch/plugins/PluginsService.java:405]: I tried to find authoritative documentation, but it's pretty confusing.

In Ruby I think the way this is disambiguated is that you go through versions and if any `~` is semver-matched then it's `true`, then you need to do an `&` for all other checks like `<=`, `>` or `!`.  This allows you to say "any 2.x, but not 2.7.1" for example.

I propose we use ^ and write tests for all these scenarios. I am fine limiting to 1 item in this array for MVP too, but it feels like we are close to the above implementation, WDYT?

Comment (abseth-amzn) [reply server/src/main/java/org/opensearch/plugins/PluginsService.java:405]: To achieve "any 2.x, but not 2.7.1", a plugin could specify two ranges, `~2.0`,  `!2.7.1` Eventually we need an `&` between them. So the logic that core should satisfy all ranges specified by the plugin should be good?

To unblock, I can validate number of ranges in the list to be one. Then I can follow up with a PR to support multiple ranges and more operators (< , > , <= , >= , !).

Comment (dblock) [reply server/src/main/java/org/opensearch/plugins/PluginsService.java:405]: You're probably right that any combination of ~ can be expressed by >= and <=, but I think it's not more code because you'd have to start raising exceptions when you have more than one.

File: server/src/main/java/org/opensearch/plugins/PluginInfo.java
Comment (abseth-amzn) [reply server/src/main/java/org/opensearch/plugins/PluginInfo.java:191]: When backporting to 2.N, the code will be updated to check for V2.N in line 187 instead of Version.V_3_0_0. 
PluginInfo serialized with V2.N-1 will only have opensearch version and not list of ranges and will be interpreted in V2.N as such in line 190. Looking to understand if the serialization issue will still persist (and in which case).

File: server/src/main/java/org/opensearch/plugins/PluginsService.java
Comment (abseth-amzn) [reply server/src/main/java/org/opensearch/plugins/PluginsService.java:405]: In case of multiple `~` ranges, the core would raise exception stating incompatible plugin because plugin wont be compatible with all of them. So we need not raise any explicit exceptions.

**Review by Bukhtawar — COMMENTED** [review]
Thanks for the changes. The one concern I see with introducing `dependencies` is that we are saying plugins could we semver compatible with dependencies outside of OpenSearch, which could easily bloat up combinations. For simplicity we could consider using opensearch version for now

**Comment by dblock** [comment]
> Thanks for the changes. The one concern I see with introducing `dependencies` is that we are saying plugins could we semver compatible with dependencies outside of OpenSearch, which could easily bloat up combinations. For simplicity we could consider using opensearch version for now

That's actually an important feature. Many plugins depend on other plugins such as job-scheduler and jar hell is already very real with 10 plugins. Allowing open dependencies will help catch dependency problems by having an easy way to codify that a plugin X is incompatible with core or plugin Y. 

@Bukhtawar I would be ok with crippling the feature for a first cut/release to only allow 1 dependency on OpenSearch if you think strongly about it, but I think we have to keep the design extensible for that scenario. WDYT?

File: server/src/main/java/org/opensearch/plugins/PluginInfo.java
Comment (saratvemulapalli) [reply server/src/main/java/org/opensearch/plugins/PluginInfo.java:191]: It shouldn't really be a problem. We constantly add new fields over the wire. 
That said this PR should have 2 additional changes:
1. Merge to main with the wire contract that semver range is in 3.x only
2. Raise another PR latest 2.x to able to read and write the new field. 
3. Update the wire compatibility check on 3.x to read the field from 2.x. 

Also lets add BWC tests, this will confirm it will work in mixed cluster and rolling upgrade cases.

**Comment by saratvemulapalli** [comment]
Love these changes. +1 to @Bukhtawar to restrict the compatible range only with OpenSearch as a dependency. Lets start small, we can eventually relax it if see the need.

Also lets add some BWC tests to make sure its wire compatible.

File: server/src/main/java/org/opensearch/plugins/PluginsService.java
Comment (dblock) [reply server/src/main/java/org/opensearch/plugins/PluginsService.java:405]: Another thought it not to implement multiple items in v1, but to implement `^3.0.0` which is another popular operator and means any 3.x (coming from https://github.com/dotnetinstaller/dotnetinstaller/issues/163#issuecomment-1891116414).

File: qa/full-cluster-restart/src/test/java/org/opensearch/upgrades/PluginInfoIT.java
Comment (peternied) [inline qa/full-cluster-restart/src/test/java/org/opensearch/upgrades/PluginInfoIT.java:25]: Nice, can we build on this assertion to check which plugins we expected to come up are available?

For the nodes that did not come online, can we do any more detailed validation to be sure why they didn't come up, I'm not clear why this test expects two nodes not to be available.

File: qa/mixed-cluster/src/test/java/org/opensearch/backwards/PluginInfoIT.java
Comment (peternied) [inline qa/mixed-cluster/src/test/java/org/opensearch/backwards/PluginInfoIT.java:30]: Can we validate the plugin version information more closely?

**Review by peternied — CHANGES_REQUESTED** [review]
Nicely done supporting the `^` operator, making great progress.

File: server/src/main/java/org/opensearch/plugins/PluginsService.java
Comment (peternied) [reply server/src/main/java/org/opensearch/plugins/PluginsService.java:405]: Thanks for adding `^` works for me!

File: qa/full-cluster-restart/src/test/java/org/opensearch/upgrades/PluginInfoIT.java
Comment (peternied) [reply qa/full-cluster-restart/src/test/java/org/opensearch/upgrades/PluginInfoIT.java:25]: @saratvemulapalli Would you mind doing a pass over these tests to see if they captured what you were looking for?

Comment (saratvemulapalli) [reply qa/full-cluster-restart/src/test/java/org/opensearch/upgrades/PluginInfoIT.java:25]: This is good, at least it says the wire compatibility works. 
Ideally the test we need is de-serialize pluginInfo on new nodes to verify we can really read them. But since we need to add some more PRs, I'd be ok to see that test in a follow up PR, since this change has been there for a while.

Also why do we see 2 nodes with restart upgrade? Shouldn't all 4 nodes come up successfully?

Comment (abseth-amzn) [reply qa/full-cluster-restart/src/test/java/org/opensearch/upgrades/PluginInfoIT.java:25]: full-cluster-restart tests use a 2 node cluster - https://github.com/opensearch-project/OpenSearch/blob/main/qa/full-cluster-restart/build.gradle#L46.

mixed-cluster tests use a 4 node cluster.

File: libs/core/src/main/java/org/opensearch/core/common/io/stream/StreamInput.java
Comment (dblock) [inline libs/core/src/main/java/org/opensearch/core/common/io/stream/StreamInput.java:755]: When we add support for multiple ranges, will this need to change or will we be adding a collection of semver ranges? Meaning does it make sense to make this a collection (of 1) right now?

**Review by dblock — COMMENTED** [review]
One question on transport, otherwise this looks good to me!

**Comment by dblock** [comment]
@reta Take a quick look in case we're missing something?

File: qa/full-cluster-restart/src/test/java/org/opensearch/upgrades/PluginInfoIT.java
Comment (peternied) [reply qa/full-cluster-restart/src/test/java/org/opensearch/upgrades/PluginInfoIT.java:25]: I'd love to read this in from a class/configuration, but this looks like a great foundation - resolving this thread.

File: server/src/main/java/org/opensearch/plugins/PluginInfo.java
Comment (peternied) [reply server/src/main/java/org/opensearch/plugins/PluginInfo.java:197]: @abseth-amzn Could create an issue to capture any follow up items such as this one for future PR - that will resolve this thread for me.

**Comment by peternied** [comment]
@abseth-amzn Could you follow up on those unresolved conversations? Really close!

File: qa/mixed-cluster/src/test/java/org/opensearch/backwards/PluginInfoIT.java
Comment (abseth-amzn) [reply qa/mixed-cluster/src/test/java/org/opensearch/backwards/PluginInfoIT.java:30]: Same test as PluginInfoIT in full-cluster-restart suite, detailed testing covered in follow up items - https://github.com/opensearch-project/OpenSearch/issues/12008

File: server/src/main/java/org/opensearch/plugins/PluginInfo.java
Comment (abseth-amzn) [reply server/src/main/java/org/opensearch/plugins/PluginInfo.java:197]: https://github.com/opensearch-project/OpenSearch/issues/12008

File: libs/core/src/main/java/org/opensearch/core/common/io/stream/StreamInput.java
Comment (abseth-amzn) [reply libs/core/src/main/java/org/opensearch/core/common/io/stream/StreamInput.java:755]: This will not need to change when we enable support for multiple ranges. The PluginInfo serialization process will write a [list of SemverRange](https://github.com/opensearch-project/OpenSearch/pull/11441/files#diff-9706884d125350e82403647f631ce066f4aa2e99747c4247f3dd943197db6fc4R211) type objects. [Every object in the list](https://github.com/opensearch-project/OpenSearch/blob/main/libs/core/src/main/java/org/opensearch/core/common/io/stream/StreamOutput.java#L702) will be [serialized as SemverType](https://github.com/opensearch-project/OpenSearch/pull/11441/files#diff-44bd9f0f10ddc587fabe125f40ee3889e5a1fa9c56b78440f41fd147df1df6bfR788) object. 
Deserialization process of PluginInfo [will read](https://github.com/opensearch-project/OpenSearch/pull/11441/files#diff-9706884d125350e82403647f631ce066f4aa2e99747c4247f3dd943197db6fc4R194) [a list of](https://github.com/opensearch-project/OpenSearch/blob/main/libs/core/src/main/java/org/opensearch/core/common/io/stream/StreamInput.java#L716) [SemverRange](https://github.com/opensearch-project/OpenSearch/blob/main/libs/core/src/main/java/org/opensearch/core/common/io/stream/StreamInput.java#L782) type objects. This has been tested in a previous revision when support for multiple ranges was added.

**Review by peternied — APPROVED** [review]

**Comment by reta** [comment]
> @reta Take a quick look in case we're missing something?

I may not be able to find the time today but please go ahead, I will review it after the merge, thanks @dblock

**Review by saratvemulapalli — APPROVED** [review]

File: server/build.gradle
Comment (reta) [inline server/build.gradle:113]: @abseth-amzn we use Jackson already, why do we need to bring GSON as well?

**Review by reta — CHANGES_REQUESTED** [review]
Folks I think we should give it a bit more time before merge

File: libs/core/src/main/java/org/opensearch/semver/SemverRange.java
Comment (reta) [inline libs/core/src/main/java/org/opensearch/semver/SemverRange.java:29]: Could we please document here what are these semver ranges and how operators should be used, similarly as we've done for `org.opensearch.bootstrap.jvm.DenyJvmVersionsParser#parse`

File: server/build.gradle
Comment (abseth-amzn) [reply server/build.gradle:113]: The server module doesn't declare a dependency on jackson or gson. I saw gson being used in a bunch of other modules, hence used it. Hoping thats ok.

File: libs/core/src/main/java/org/opensearch/semver/SemverRange.java
Comment (abseth-amzn) [reply libs/core/src/main/java/org/opensearch/semver/SemverRange.java:35]: Updated javadoc

**Comment by abseth-amzn** [comment]
> Folks I think we should give it a bit more time before merge

@reta - Thank you for the feedback. I have updated the javadoc and responded to the dependency comment. It would be helpful if you could share any concerns you may have or any thing that needs attention.

File: server/build.gradle
Comment (reta) [reply server/build.gradle:113]: It does not declare the dependency explicitly but it is there, coming from `opensearch-x-content`, `openseach-core`, the good part - it is coming in `lib/` folder already so we are not risking to break anyone with this new dependency (by JARhelling), thank you.

Comment (abseth-amzn) [reply server/build.gradle:113]: Thanks @reta. Those modules are bringing in jackson-core and jackson-dataformat, but not jackson-databind which would be needed here.
Also, jackson does not allow unquoted field values, so following specifications wont be allowed as quotes would be required around the version and operator:
```
dependencies={ opensearch: x.y.z }
dependencies={ opensearch: =x.y.z }
dependencies={ opensearch: ~x.y.z }
```
gson allows for some flexibility here.

Comment (reta) [reply server/build.gradle:113]: > gson allows for some flexibility here.

Does it justify the need for new dependencies? The Jackson allow to read unquoted  field names, I think it is an easy ask to require values to be always quoted, since in many cases they have to be quoited anyway:

```
"{opensearch:"~1.2.3, =1.2.3"}
```

Comment (abseth-amzn) [reply server/build.gradle]: @reta - Updated to use jackson instead of gson.

File: server/src/main/java/org/opensearch/plugins/PluginInfo.java
Comment (reta) [inline server/src/main/java/org/opensearch/plugins/PluginInfo.java:281]: You don't need to bring whole `jackson-databind` here, the `jackson-core` + XContent parsing we already have (as we discussed) is fully sufficient:

```
           // This could be made a static / final instance, no need to create every time
            final JsonFactory jsonFactory = new JsonFactory();
            jsonFactory.configure(JsonReadFeature.ALLOW_UNQUOTED_FIELD_NAMES.mappedFeature(), true);
            
            Map<String, Object> dependenciesMap = null;
            try (final JsonXContentParser parser = new JsonXContentParser(NamedXContentRegistry.EMPTY, DeprecationHandler.IGNORE_DEPRECATIONS, jsonFactory.createParser(dependenciesValue))) {
                dependenciesMap = parser.map(); 
            }
```

File: server/src/main/java/org/opensearch/plugins/PluginInfo.java
Comment (reta) [inline server/src/main/java/org/opensearch/plugins/PluginInfo.java:76]:
```suggestion
    private static final JsonFactory jsonFactory = new JsonFactory().
        configure(JsonReadFeature.ALLOW_UNQUOTED_FIELD_NAMES.mappedFeature(), true);;
```

File: server/src/main/java/org/opensearch/plugins/PluginInfo.java
Comment (reta) [inline server/src/main/java/org/opensearch/plugins/PluginInfo.java:88]: I don't think you need this block, `JsonFactory` could be initialized in one shot

File: server/build.gradle
Comment (reta) [inline server/build.gradle:71]: Accidental?

**Comment by reta** [comment]
@abseth-amzn a few minor comments but LGTM otherwise, thank you. The DCO check is failing

File: server/src/main/java/org/opensearch/plugins/PluginInfo.java
Comment (abseth-amzn) [reply server/src/main/java/org/opensearch/plugins/PluginInfo.java]: Updated

File: server/src/main/java/org/opensearch/plugins/PluginInfo.java
Comment (abseth-amzn) [reply server/src/main/java/org/opensearch/plugins/PluginInfo.java]: Updated based on recommendation.

**Review by reta — APPROVED** [review]

**Comment by reta** [comment]
@abseth-amzn we still have DCO failures sadly for some commits, may need rebase + squash

**Comment by rursprung** [comment]
thanks, this should simplify upgrades greatly!
will this be cherry-picked to the 2.12 or at least the 2.x branch?

**Comment by abseth-amzn** [comment]
> thanks, this should simplify upgrades greatly! will this be cherry-picked to the 2.12 or at least the 2.x branch?

Yes, working on backporting it to 2.x branch.

**Comment by abseth-amzn** [comment]
Backport PR - https://github.com/opensearch-project/OpenSearch/pull/12271

**Comment by dblock** [comment]
@rursprung AFAIK no. If you are looking to a semi-production plugin to contribute an example to, may I volunteer https://github.com/dblock/opensearch-api? :)

**Comment by rursprung** [comment]
> @rursprung AFAIK no. If you are looking to a semi-production plugin to contribute an example to, may I volunteer https://github.com/dblock/opensearch-api? :)

i wanted to contribute it to another plugin but i fail to see how it can be done with the gradle plugin - either i'm missing something or it's not possible yet with that one? 😕

**Comment by dblock** [comment]
@rursprung I don't think you're missing anything, probably not implemented

**Comment by rursprung** [comment]
i've raised #13187 to address this

**Pull request merged**
